src/formatting.rs

```diff
@@ -73,7 +73,12 @@ fn format_project<T: FormatHandler>(
     let source_map = Rc::new(SourceMap::new(FilePathMapping::empty()));
     let mut parse_session = make_parse_sess(source_map.clone(), config);
     let mut report = FormatReport::new();
-    let krate = parse_crate(input, &parse_session, config, &mut report)?;
+    let krate = match parse_crate(input, &parse_session, config, &mut report) {
+        Ok(krate) => krate,
+        // Surface parse error via Session (errors are merged there from report)
+        Err(ErrorKind::ParseError) => return Ok(report),
```

Comment:
Changed this to return report here for parse errors, otherwise we eagerly `?`-return an `Err` here:
https://github.com/rust-lang/rustfmt/blob/66c98db50f58aa97be7bdd34af136ad258c6e04c/src/formatting.rs#L76
but we only add errors to `Session` if we have `Ok(FormatReport)`
https://github.com/rust-lang/rustfmt/blob/66c98db50f58aa97be7bdd34af136ad258c6e04c/src/formatting.rs#L49-L57

I changed this because I expected `session.has_parsing_errors()` call to work:
https://github.com/rust-lang/rustfmt/pull/3240/files#diff-382d140f826c9bd99f62116b9712fad3R299
and it seemed like a special case that it didn't.

However, for IO errors we still will get `Err(_)` from `Session::format(...)` and `session.has_operational_errors()` will return `false` (should we also change that?).